🌲🪟Fixes to the remote inspection LLVM headers

etcwilde opened the pull request:
This patch fixes the location of the llvm remote inspection headers for
    MSVC header lookup. MSVC appears to search in the directory of the
    current header before returning to the specified header search
    directories. When building SwiftRemoteMirror, the file contains a
    reference to `swift/RemoteInspection/ReflectionContext.h`. Under
    RelfectionContext.h, there is an include of "llvm/BinaryFormat/COFF.h".
    Because there is an `llvm` and `llvm-c` directory inside of
    `swift/RemoteInspection/`, and `ReflectionContext.h` is in that
    directory, MSVC is expanding the `COFF.h` inside of the
    RemoteInspection headers instead of the copy in LLVM itself, resulting
    in eventually finding usages of `countPopulation` instead of using the
    new `llvm::popcount` API, ultimately resulting in a build failure.
    
   The fix is to ensure that the `llvm` header directory does not live
    immediately next to the headers in RemoteInspection, but instead offset
    them by one. The LLVM headers copied into RemoteInspection are supposed
    to be used when compiling the runtime libraries, so I chose the name
    "RuntimeHeaders".

Note, there are some ODR issues here if someone were to try, say, implementing their own ABI for RemoteInspection, but that would still be an issue today and this is needed for rebranch work.

rdar://114413755

etcwilde commented:
I am not really sure how to fix the ODR stuff in the C headers... so hopefully that's fine. :D

compnerd commented on include/swift/RemoteInspection/RuntimeHeaders/llvm/ADT/BitmaskEnum.h:
```diff
@@ -89,6 +89,8 @@
   using ::llvm::BitmaskEnumDetail::operator^=
 
+inline namespace __swift {
+inline namespace __runtime {
```
This seems like playing with different ODR.  This is the namespace that we use in the runtime, so that would now cause a possible issue.  Can we use something like `__swift::__inspection`?

etcwilde replied:
Done

etcwilde commented:
@swift-ci please smoke test

etcwilde commented:
Linux failure:

```
/home/build-user/swift/include/swift/RemoteInspection/RuntimeHeaders/llvm/Support/ErrorOr.h:257:5: error: no template named 'AlignedCharArrayUnion'; did you mean '::llvm::AlignedCharArrayUnion'?
  257 |     AlignedCharArrayUnion<storage_type> TStorage;
      |     ^~~~~~~~~~~~~~~~~~~~~
      |     ::llvm::AlignedCharArrayUnion
```

Introduced new build failures to Linux build.

compnerd commented:
I think that is expected if there are uses of the types which are not marked as being in the namespace.

etcwilde committed:
Fix RemoteInspection LLVM header location

This patch fixes the location of the llvm remote inspection headers for
MSVC header lookup. MSVC appears to search in the directory of the
current header before returning to the specified header search
directories. When building SwiftRemoteMirror, the file contains a
reference to `swift/RemoteInspection/ReflectionContext.h`. Under
RelfectionContext.h, there is an include of "llvm/BinaryFormat/COFF.h".
Because there is an `llvm` and `llvm-c` directory inside of
`swift/RemoteInspection/`, and `ReflectionContext.h` is in that
directory, MSVC is expanding the `COFF.h` inside of the
RemoteInspection headers instead of the copy in LLVM itself, resulting
in eventually finding usages of `countPopulation` instead of using the
new `llvm::popcount` API, ultimately resulting in a build failure.

The fix is to ensure that the `llvm` header directory does not live
immediately next to the headers in RemoteInspection, but instead offset
them by one. The LLVM headers copied into RemoteInspection are supposed
to be used when compiling the runtime libraries, so I chose the name
"RuntimeHeaders".

etcwilde commented:
@swift-ci please smoke test

etcwilde commented:
Windows still has the old SDK, so no meaningful output. Local Windows is getting much further now though.

Just making sure I didn't regress Linux and macOS now.
